File: backend/src/main/java/com/example/nd/controller/FileController.java
package com.example.nd.controller;

import com.example.nd.dto.*;
import com.example.nd.mapper.FileMetadataMapper;
import com.example.nd.model.File;
import com.example.nd.model.FileInfo;
import com.example.nd.model.UploadTask;
import com.example.nd.service.CoverService;
import com.example.nd.service.FileSearchService;
import com.example.nd.service.FileService;
import com.example.nd.service.UploadService;
import com.example.nd.util.AuthUtil;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.io.Resource;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.List;

@RestController
@RequestMapping("/files")
@Tag(name = "文件管理", description = "文件操作相关接口")
public class FileController {

    @Autowired
    private FileService fileService;

    @Autowired
    private UploadService uploadService;
    
    @Autowired
    private CoverService coverService;
    
    @Autowired
    private FileMetadataMapper fileMetadataMapper;
    
    @Autowired
    private FileSearchService fileSearchService;


    @GetMapping
    @Operation(summary = "获取文件列表", description = "根据文件夹ID获取文件列表")
    public ApiResponse<List<FileInfo>> getFilesByFolderId(@RequestParam(required = false) Long folderId) {
        Long userId = AuthUtil.getUserId();
        List<FileInfo> files;
        if (folderId == null || folderId == 0) {
            files = fileService.getRootFiles(userId);
        } else {
            files = fileService.getFilesByFolderId(folderId);
        }
        return ApiResponse.success(files);
    }

    @GetMapping("/path")
    @Operation(summary = "获取文件夹路径", description = "根据文件夹ID获取从根目录到该文件夹的完整路径")
    public ApiResponse<List<FileInfo>> getFolderPath(@RequestParam(required = false) Long folderId) {
        Long userId = AuthUtil.getUserId();
        List<FileInfo> path = fileService.getFolderPath(folderId, userId);
        return ApiResponse.success(path);
    }

    @GetMapping("/search")
    @Operation(summary = "搜索文件", description = "根据关键词、文件类型、大小、日期等条件搜索文件")
    public ApiResponse<List<SearchResult>> searchFiles(@RequestParam(required = false) String keyword,
                                                     @RequestParam(required = false) String fileType,
                                                     @RequestParam(required = false) Long minSize,
                                                     @RequestParam(required = false) Long maxSize,
                                                     @RequestParam(required = false) Long folderId,
                                                     @RequestParam(required = false) String startDate,
                                                     @RequestParam(required = false) String endDate,
                                                     @RequestParam(required = false) String sortBy,
                                                     @RequestParam(required = false) String sortOrder,
                                                     @RequestParam(required = false, defaultValue = "0") Integer page,
                                                     @RequestParam(required = false, defaultValue = "20") Integer pageSize) {
        Long userId = AuthUtil.getUserId();
        SearchRequest request = new SearchRequest();
        request.setKeyword(keyword);
        request.setFileType(fileType);
        request.setMinSize(minSize);
        request.setMaxSize(maxSize);
        request.setFolderId(folderId);
        
        if (startDate != null && !startDate.isEmpty()) {
            request.setStartDate(java.time.LocalDateTime.parse(startDate));
        }
        if (endDate != null && !endDate.isEmpty()) {
            request.setEndDate(java.time.LocalDateTime.parse(endDate));
        }
        
        request.setSortBy(sortBy);
        request.setSortOrder(sortOrder);
        request.setPage(page);
        request.setPageSize(pageSize);
        request.setOffset(page * pageSize);
        
        List<SearchResult> results = fileSearchService.searchFiles(userId, request);
        return ApiResponse.success(results);
    }

    @PostMapping("/upload")
    @Operation(summary = "上传文件", description = "上传单个文件")
    public ApiResponse<FileInfo> uploadFile(@RequestParam("file") MultipartFile file, @RequestParam(required = false) Long parentFolderId) {
        Long userId = AuthUtil.getUserId();
        FileInfo fileInfo = fileService.uploadFile(userId, file, parentFolderId);
        return ApiResponse.success(fileInfo);
    }

    @GetMapping("/{fileId}")
    @Operation(summary = "获取文件信息", description = "根据文件ID获取文件详细信息")
    public ApiResponse<FileInfo> getFileById(@PathVariable Long fileId) {
        Long userId = AuthUtil.getUserId();
        FileInfo fileInfo = fileService.getFileById(fileId);
        if (fileInfo != null) {
            return ApiResponse.success(fileInfo);
        }
        return ApiResponse.error("File not found");
    }

    @GetMapping("/{fileId}/download")
    @Operation(summary = "下载文件", description = "下载指定文件")
    public ResponseEntity<Resource> downloadFile(@PathVariable Long fileId) {
        Long userId = AuthUtil.getUserId();
        FileInfo file = fileService.getFileById(fileId);
        if (file != null && file.getDeletedAt() == null) {
            try {
                Path filePath = Paths.get(file.getStoragePath());
                Resource resource = new org.springframework.core.io.UrlResource(filePath.toUri());

                return ResponseEntity.ok()
                        .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"" + file.getOriginalName() + "\"")
                        .contentType(MediaType.parseMediaType(file.getMimeType()))
                        .body(resource);
            } catch (IOException e) {
                return ResponseEntity.internalServerError().build();
            }
        }
        return ResponseEntity.notFound().build();
    }

    @PutMapping("/{fileId}/move")
    @Operation(summary = "移动文件", description = "将文件移动到指定文件夹")
    public ApiResponse<String> moveFile(@PathVariable Long fileId, @RequestBody FileMoveRequest request) {
        Long userId = AuthUtil.getUserId();
        fileService.moveFile(fileId, request.getTargetFolderId());
        return ApiResponse.success("File moved successfully");
    }

    @PutMapping("/{fileId}/rename")
    @Operation(summary = "重命名文件", description = "重命名文件")
    public ApiResponse<String> renameFile(@PathVariable Long fileId, @RequestBody FileRenameRequest request) {
        Long userId = AuthUtil.getUserId();
        fileService.renameFile(fileId, request.getNewName());
        return ApiResponse.success("File renamed successfully");
    }

    @DeleteMapping("/{fileId}")
    @Operation(summary = "删除文件", description = "删除指定文件")
    public ApiResponse<String> deleteFile(@PathVariable Long fileId) {
        Long userId = AuthUtil.getUserId();
        fileService.deleteFile(fileId);
        return ApiResponse.success("File deleted successfully");
    }

    @PostMapping("/folders")
    @Operation(summary = "创建文件夹", description = "创建新文件夹")
    public ApiResponse<FileInfo> createFolder(@RequestBody FolderCreateRequest request) {
        Long userId = AuthUtil.getUserId();
        FileInfo folder = fileService.createFolder(userId, request.getFolderName(), request.getParentFolderId());
        return ApiResponse.success(folder);
    }

    @PostMapping("/check")
    @Operation(summary = "秒传检查", description = "检查文件是否已存在")
    public ApiResponse<CheckFileResponse> checkFileExist(@RequestBody CheckFileRequest request) {
        CheckFileResponse response = uploadService.checkFileExist(request);
        return ApiResponse.success(response);
    }

    @PostMapping("/upload/init")
    @Operation(summary = "上传初始化", description = "初始化上传任务")
    public ApiResponse<UploadInitResponse> initializeUpload(@RequestBody UploadInitRequest request) {
        UploadInitResponse response = uploadService.initializeUpload(request);
        return ApiResponse.success(response);
    }

    @PostMapping("/upload/chunk")
    @Operation(summary = "分片上传", description = "上传文件分片")
    public ApiResponse<String> uploadChunk(
            @RequestParam("uploadId") String uploadId,
            @RequestParam("chunkIndex") Integer chunkIndex,
            @RequestParam("chunkData") MultipartFile chunkData) {
        uploadService.uploadChunk(uploadId, chunkIndex, chunkData);
        return ApiResponse.success("Chunk uploaded successfully");
    }

    @PostMapping("/upload/complete")
    @Operation(summary = "完成上传", description = "完成上传，合并分片")
    public ApiResponse<FileInfo> completeUpload(@RequestBody UploadCompleteRequest request) {
        File file = uploadService.completeUpload(request);
        FileInfo fileInfo = new FileInfo();
        fileInfo.setId(file.getId());
        fileInfo.setUserId(file.getUserId());
        fileInfo.setParentFolderId(file.getParentId());
        fileInfo.setFileName(file.getName());
        fileInfo.setOriginalName(file.getName());
        fileInfo.setCreatedAt(file.getCreatedAt());
        fileInfo.setUpdatedAt(file.getUpdatedAt());
        fileInfo.setDeletedAt(file.getDeletedAt());
        fileInfo.setIsFolder(file.getIsFolder());
        return ApiResponse.success(fileInfo);
    }

    @GetMapping("/upload/status/{uploadId}")
    @Operation(summary = "获取上传状态", description = "获取上传状态")
    public ApiResponse<UploadTask> getUploadStatus(@PathVariable String uploadId) {
        UploadTask uploadTask = uploadService.getUploadStatus(uploadId);
        return ApiResponse.success(uploadTask);
    }

    @PostMapping("/upload/cancel/{uploadId}")
    @Operation(summary = "取消上传", description = "取消上传任务")
    public ApiResponse<String> cancelUpload(@PathVariable String uploadId) {
        uploadService.cancelUpload(uploadId);
        return ApiResponse.success("Upload cancelled successfully");
    }
    
    @GetMapping("/{fileId}/cover")
    @Operation(summary = "获取文件封面", description = "获取文件的封面图片")
    public ResponseEntity<Resource> getFileCover(@PathVariable Long fileId, @RequestParam(required = false) String size) {
        try {
            Path coverPath = coverService.getCover(fileId, size);
            if (coverPath != null && Files.exists(coverPath)) {
                Resource resource = new org.springframework.core.io.UrlResource(coverPath.toUri());
                return ResponseEntity.ok()
                        .header(HttpHeaders.CONTENT_DISPOSITION, "inline; filename=\"cover.jpg\"")
                        .contentType(MediaType.IMAGE_JPEG)
                        .body(resource);
            }
            return ResponseEntity.notFound().build();
        } catch (Exception e) {
            return ResponseEntity.internalServerError().build();
        }
    }
}

File: backend_python/app/controllers/file_controller.py
from fastapi import (
    APIRouter,
    Depends,
    HTTPException,
    status,
    UploadFile,
    File,
    Form,
    Request,
)
from sqlalchemy.orm import Session
from app.database import get_db
from app.schemas import (
    FileInfo,
    FileMoveRequest,
    FileRenameRequest,
    FolderCreateRequest,
    CheckFileRequest,
    CheckFileResponse,
    UploadInitRequest,
    UploadInitResponse,
    UploadCompleteRequest,
    SearchResult,
    SearchRequest,
    ApiResponse,
)
from app.services.file_service import FileService
from app.services.share_service import ShareService
from app.dependencies import get_current_active_user
from app.models import User
from typing import Optional, List
import os

router = APIRouter(prefix="/files", tags=["文件管理"])


@router.get("", response_model=dict)
def get_files_by_folder_id(
    folder_id: Optional[int] = None,
    current_user: User = Depends(get_current_active_user),
):
    """获取文件列表"""
    try:
        file_service = FileService(current_user.__dict__.get("_db"))

        if folder_id is None or folder_id == 0:
            files = file_service.get_root_files(current_user.id)
        else:
            files = file_service.get_files_by_folder_id(folder_id)

        return {
            "code": 200,
            "message": "success",
            "data": [file.dict() for file in files],
        }
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get files: {str(e)}",
        )


@router.get("/path", response_model=dict)
def get_folder_path(
    folder_id: Optional[int] = None,
    current_user: User = Depends(get_current_active_user),
):
    """获取文件夹路径"""
    try:
        file_service = FileService(current_user.__dict__.get("_db"))
        path = file_service.get_folder_path(folder_id, current_user.id)

        return {
            "code": 200,
            "message": "success",
            "data": [file.dict() for file in path],
        }
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get folder path: {str(e)}",
        )


@router.get("/search", response_model=dict)
def search_files(
    keyword: Optional[str] = None,
    file_type: Optional[str] = None,
    min_size: Optional[int] = None,
    max_size: Optional[int] = None,
    folder_id: Optional[int] = None,
    start_date: Optional[str] = None,
    end_date: Optional[str] = None,
    sort_by: Optional[str] = None,
    sort_order: Optional[str] = "desc",
    page: int = 0,
    page_size: int = 20,
    current_user: User = Depends(get_current_active_user),
):
    """搜索文件"""
    try:
        file_service = FileService(current_user.__dict__.get("_db"))

        # 处理日期字符串
        from datetime import datetime

        start_datetime = None
        end_datetime = None

        if start_date:
            start_datetime = datetime.fromisoformat(start_date.replace("Z", "+00:00"))
        if end_date:
            end_datetime = datetime.fromisoformat(end_date.replace("Z", "+00:00"))

        search_request = SearchRequest(
            keyword=keyword,
            file_type=file_type,
            min_size=min_size,
            max_size=max_size,
            folder_id=folder_id,
            start_date=start_datetime,
            end_date=end_datetime,
            sort_by=sort_by,
            sort_order=sort_order,
            page=page,
            page_size=page_size,
        )

        results = file_service.search_files(current_user.id, search_request)

        return {
            "code": 200,
            "message": "success",
            "data": [result.dict() for result in results],
        }
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to search files: {str(e)}",
        )


@router.post("/upload", response_model=dict)
async def upload_file(
    file: UploadFile = File(...),
    parent_folder_id: Optional[int] = Form(None),
    current_user: User = Depends(get_current_active_user),
):
    """上传文件"""
    try:
        file_service = FileService(current_user.__dict__.get("_db"))
        uploaded_file = await file_service.upload_file(
            current_user.id, file, parent_folder_id
        )

        return {"code": 200, "message": "success", "data": uploaded_file.dict()}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to upload file: {str(e)}",
        )


@router.get("/{file_id}", response_model=dict)
def get_file_by_id(file_id: int, current_user: User = Depends(get_current_active_user)):
    """获取文件信息"""
    try:
        file_service = FileService(current_user.__dict__.get("_db"))
        file_info = file_service.get_file_by_id(file_id)

        if not file_info:
            return {"code": 404, "message": "File not found", "data": None}

        return {"code": 200, "message": "success", "data": file_info.dict()}
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get file: {str(e)}",
        )


@router.get("/{file_id}/download")
async def download_file(
    file_id: int, current_user: User = Depends(get_current_active_user)
):
    """下载文件"""
    try:
        file_service = FileService(current_user.__dict__.get("_db"))
        file_info = file_service.get_file_by_id(file_id)

        if not file_info:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND, detail="File not found"
            )

        if file_info.is_folder:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST, detail="Cannot download folder"
            )

        # 返回文件路径，实际应用中应该返回文件流
        return {
            "file_path": file_info.storage_path,
            "file_name": file_info.file_name,
            "file_size": file_info.file_size,
            "mime_type": file_info.mime_type,
        }
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to download file: {str(e)}",
        )


@router.put("/{file_id}/move", response_model=dict)
def move_file(
    file_id: int,
    request: FileMoveRequest,
    current_user: User = Depends(get_current_active_user),
):
    """移动文件"""
    try:
        file_service = FileService(current_user.__dict__.get("_db"))
        file_service.move_file(file_id, request.target_folder_id)

        return {"code": 200, "message": "success", "data": "File moved successfully"}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to move file: {str(e)}",
        )


@router.put("/{file_id}/rename", response_model=dict)
def rename_file(
    file_id: int,
    request: FileRenameRequest,
    current_user: User = Depends(get_current_active_user),
):
    """重命名文件"""
    try:
        file_service = FileService(current_user.__dict__.get("_db"))
        file_service.rename_file(file_id, request.new_name)

        return {"code": 200, "message": "success", "data": "File renamed successfully"}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to rename file: {str(e)}",
        )


@router.delete("/{file_id}", response_model=dict)
def delete_file(file_id: int, current_user: User = Depends(get_current_active_user)):
    """删除文件"""
    try:
        file_service = FileService(current_user.__dict__.get("_db"))
        file_service.delete_file(file_id)

        return {"code": 200, "message": "success", "data": "File deleted successfully"}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to delete file: {str(e)}",
        )


@router.post("/folders", response_model=dict)
def create_folder(
    request: FolderCreateRequest, current_user: User = Depends(get_current_active_user)
):
    """创建文件夹"""
    try:
        file_service = FileService(current_user.__dict__.get("_db"))
        folder = file_service.create_folder(
            current_user.id, request.folder_name, request.parent_folder_id
        )

        return {"code": 200, "message": "success", "data": folder.dict()}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to create folder: {str(e)}",
        )


@router.post("/check", response_model=dict)
def check_file_exist(
    request: CheckFileRequest, current_user: User = Depends(get_current_active_user)
):
    """秒传检查"""
    try:
        file_service = FileService(current_user.__dict__.get("_db"))
        result = file_service.check_file_exist(request)

        return {"code": 200, "message": "success", "data": result.dict()}
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to check file: {str(e)}",
        )


@router.post("/upload/init", response_model=dict)
def initialize_upload(
    request: UploadInitRequest, current_user: User = Depends(get_current_active_user)
):
    """上传初始化"""
    try:
        file_service = FileService(current_user.__dict__.get("_db"))
        result = file_service.initialize_upload(request)

        return {"code": 200, "message": "success", "data": result.dict()}
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to initialize upload: {str(e)}",
        )


@router.post("/upload/chunk", response_model=dict)
async def upload_chunk(
    upload_id: str,
    chunk_index: int,
    chunk_data: UploadFile = File(...),
    current_user: User = Depends(get_current_active_user),
):
    """分片上传"""
    try:
        # 这里应该实现分片上传逻辑
        return {
            "code": 200,
            "message": "success",
            "data": "Chunk uploaded successfully",
        }
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to upload chunk: {str(e)}",
        )


@router.post("/upload/complete", response_model=dict)
def complete_upload(
    request: UploadCompleteRequest,
    current_user: User = Depends(get_current_active_user),
):
    """完成上传"""
    try:
        file_service = FileService(current_user.__dict__.get("_db"))
        file = file_service.complete_upload(request)

        return {"code": 200, "message": "success", "data": file.dict()}
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to complete upload: {str(e)}",
        )


@router.get("/upload/status/{upload_id}", response_model=dict)
def get_upload_status(
    upload_id: str, current_user: User = Depends(get_current_active_user)
):
    """获取上传状态"""
    try:
        # 这里应该查询上传任务状态
        return {
            "code": 200,
            "message": "success",
            "data": {"upload_id": upload_id, "status": "completed"},
        }
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get upload status: {str(e)}",
        )


@router.post("/upload/cancel/{upload_id}", response_model=dict)
def cancel_upload(
    upload_id: str, current_user: User = Depends(get_current_active_user)
):
    """取消上传"""
    try:
        # 这里应该取消上传任务
        return {
            "code": 200,
            "message": "success",
            "data": "Upload cancelled successfully",
        }
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to cancel upload: {str(e)}",
        )
